### PR TITLE
refactor: improved NameErrors and NonExhaustiveMatchError.

### PR DESCRIPTION
**Pull request by magnus-madsen**
Related to #2691

### CODE REVIEW

**Comment by magnus-madsen** [comment]
This is my first attempt to improve error messages. You will note that I did not fill in every single `explain` text. This is either because I think it might not be necessary (for now) or because I simple did not know what to say.

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 1]
```diff
@@ -38,12 +38,12 @@ object NameError {
...
     */
   case class AmbiguousVarOrUse(name: String, loc: SourceLocation, loc1: SourceLocation, loc2: SourceLocation) extends NameError {
-    def summary: String = s"Ambiguous name. The name may refer to both a variable and a use."
+    def summary: String = s"Ambiguous name: '$name'. The name may refer to both a variable and a use of a name."
 
     def message(formatter: Formatter): String = {
       import formatter._
       s"""${line(kind, source.format)}
-         |>> Ambiguous name '${red(name)}'. The name may refer to both a variable and a use.
+         |>> Ambiguous name '${red(name)}'. The name may refer to both a variable and a use of a name.
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:46]: Would it be possible to show both occurrences?

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 2]
```diff
@@ -69,25 +70,29 @@ object NameError {
...
+         |${code(loc2, "the second definition was here.")}
          |""".stripMargin
     }
 
-    def loc: SourceLocation = loc1
-
     def explain(formatter: Formatter): Option[String] = Some({
-      import formatter._
-      s"${underline("Tip:")} Remove or rename one of the occurrences."
+      """Note: Flix does not support function overloading, i.e. you cannot define two
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:87]: Maybe the "Note:" should be removed as it's not present in any other explanation?

**Review by jaschdoc — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 3]
```diff
@@ -214,7 +207,7 @@ object NameError {
     * @param loc  the location of the suspicious type variable.
     */
   case class SuspiciousTypeVarName(name: String, loc: SourceLocation) extends NameError {
-    def summary: String = s"Suspicious type variable. Did you mean: '${name.capitalize}'?"
+    def summary: String = s"Suspicious type variable '$name'. Did you mean: '${name.capitalize}'?"
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:210]: `int` is in the list of suspicious names, but `Int` is not a valid type anymore (or at least we are headed in that direction).

So we may just want to take two parameters here, one for the suspicious name and one for our guess.

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 4]
```diff
@@ -239,21 +237,20 @@ object NameError {
...
+         |${code(loc, "undefined Java class.")}
          |""".stripMargin
     }
 
-    /**
-      * Returns a formatted string with helpful suggestions.
-      */
-    def explain(formatter: Formatter): Option[String] = None
+    def explain(formatter: Formatter): Option[String] = Some({
+      "Flix cannot find the Java class. Maybe there is a typo?"
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:252]: I think there is also a decent chance that the user forgot to include the jar when compiling.

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 5]
```diff
@@ -313,20 +307,20 @@ object NameError {
...
-         |${code(loc, "Illegal signature.")}
+         |${code(loc, "unexpected signature.")}
          |""".stripMargin
     }
 
     def explain(formatter: Formatter): Option[String] = Some({
-      import formatter._
-      s"${underline("Tip:")} Change the signature to include the class type parameter, or remove the signature."
+      """Every signature in a type class must mention the type variable of the class.
+        |""".stripMargin
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:323]: Nitpick: some of these are single-line and some end with a newline.

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 6]
```diff
@@ -69,25 +70,29 @@ object NameError {
...
-    def loc: SourceLocation = loc1
-
     def explain(formatter: Formatter): Option[String] = Some({
-      import formatter._
-      s"${underline("Tip:")} Remove or rename one of the occurrences."
+      """Note: Flix does not support function overloading, i.e. you cannot define two
+        |functions with the same name, even if their formal parameters differ.
+        |
+        |If you want two functions to share the same name you have to either (a) put them
+        |into separate namespaces or (b) to use a type class.
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:91]: IMO it's more readable to put these options on their own lines

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala [hunk 7]
```diff
@@ -226,8 +219,13 @@ object NameError {
     }
 
     def explain(formatter: Formatter): Option[String] = Some({
-      import formatter._
-      s"${underline("Tip:")} Type variables are always lowercase. Named types are uppercase."
+      """In Flix every type variable begins with a lowercase name. Your type variable
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/NameError.scala:222]: Another nitpick: everywhere else, `Flix` is kind of personified: "Flix requires XYZ". Here we have a slightly different style where we look at Flix more from the outside.

**Comment by magnus-madsen** [comment]
Thanks. Ill try to incorporate your changes!

File: main/src/ca/uwaterloo/flix/language/errors/NameError.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/errors/NameError.scala:46]: it does

**Pull request merged**